process/track/blockProcessor.go

```diff
@@ -0,0 +1,265 @@
+	return lastCrossNotarizedHeader, lastCrossNotarizedHeaderHash, headers, hashes
+}
+
+func (bp *blockProcessor) computeSelfNotarizedHeaders(headers []data.HeaderHandler) ([]data.HeaderHandler, [][]byte) {
+	selfNotarizedHeadersInfo := make([]*headerInfo, 0)
+
+	for _, header := range headers {
+		selfHeadersInfo := bp.blockTracker.getSelfHeaders(header)
+		if len(selfHeadersInfo) > 0 {
+			selfNotarizedHeadersInfo = append(selfNotarizedHeadersInfo, selfHeadersInfo...)
+		}
+	}
+
+	if len(selfNotarizedHeadersInfo) > 1 {
+		sort.Slice(selfNotarizedHeadersInfo, func(i, j int) bool {
+			return selfNotarizedHeadersInfo[i].header.GetNonce() < selfNotarizedHeadersInfo[j].header.GetNonce()
+		})
+	}
+
+	selfNotarizedHeaders := make([]data.HeaderHandler, 0)
+	selfNotarizedHeadersHashes := make([][]byte, 0)
+
+	for _, selfNotarizedHeaderInfo := range selfNotarizedHeadersInfo {
+		selfNotarizedHeaders = append(selfNotarizedHeaders, selfNotarizedHeaderInfo.header)
+		selfNotarizedHeadersHashes = append(selfNotarizedHeadersHashes, selfNotarizedHeaderInfo.hash)
+	}
+
+	return selfNotarizedHeaders, selfNotarizedHeadersHashes
+}
+
+func (bp *blockProcessor) computeLongestChain(shardID uint32, header data.HeaderHandler) ([]data.HeaderHandler, [][]byte) {
+	headers := make([]data.HeaderHandler, 0)
+	headersHashes := make([][]byte, 0)
+
+	if check.IfNil(header) {
+		return headers, headersHashes
+	}
+
+	sortedHeaders, sortedHeadersHashes := bp.blockTracker.sortHeadersFromNonce(shardID, header.GetNonce()+1)
+	if len(sortedHeaders) == 0 {
+		return headers, headersHashes
+	}
+
+	longestChainHeadersIndexes := make([]int, 0)
+	headersIndexes := make([]int, 0)
+	bp.getNextHeader(&longestChainHeadersIndexes, headersIndexes, header, sortedHeaders, 0)
+
+	for _, index := range longestChainHeadersIndexes {
+		headers = append(headers, sortedHeaders[index])
+		headersHashes = append(headersHashes, sortedHeadersHashes[index])
+	}
+
+	return headers, headersHashes
+}
+
+func (bp *blockProcessor) getNextHeader(
+	longestChainHeadersIndexes *[]int,
+	headersIndexes []int,
+	prevHeader data.HeaderHandler,
+	sortedHeaders []data.HeaderHandler,
+	index int,
+) {
+	defer func() {
+		if len(headersIndexes) > len(*longestChainHeadersIndexes) {
+			*longestChainHeadersIndexes = headersIndexes
+		}
+	}()
+
+	if check.IfNil(prevHeader) {
+		return
+	}
+
+	for i := index; i < len(sortedHeaders); i++ {
+		currHeader := sortedHeaders[i]
+		if currHeader.GetNonce() > prevHeader.GetNonce()+1 {
+			break
+		}
+
+		if currHeader.GetNonce() == prevHeader.GetNonce()+1 {
+			err := bp.headerValidator.IsHeaderConstructionValid(currHeader, prevHeader)
+			if err != nil {
+				continue
+			}
+
+			err = bp.checkHeaderFinality(currHeader, sortedHeaders, i+1)
+			if err != nil {
+				continue
+			}
+
+			headersIndexes = append(headersIndexes, i)
+			bp.getNextHeader(longestChainHeadersIndexes, headersIndexes, currHeader, sortedHeaders, i+1)
```

Comment:
i think it would be better to do it iteratively: https://stackoverflow.com/questions/15013956/height-of-binary-search-tree-iteratively/15015738

Reply:
This method from the link is good if you want to determine only the height of the trie. In my case I need to returns the real value from each node in the longest chain (slice of headers) so it would be much complicated as I need to use additional lists. More than this, for our case the method from the link would consumes much time as it would use a lot of deletes from the lists. In our case when there could not be more than one chain with many children and the rest of the nodes from this trie would have no children (because there could be hardly more than one fork) would not make sense to change this approach. Its for sure that it would be calculated near than instantly as this computation is done only upward from the last final and all the records behind final would be deleted.